### PR TITLE
Add form commands

### PR DESCRIPTION
**Pull request by tothandras**
Usage:

```sh
➜ banzai form
Open forms from config, persist provided values and generate templates

Usage:
  banzai form [command]

Available Commands:
  migrate     Migrate form values from source config to target config
  open        Open form defined in the form config file
  template    Execute form template(s)

➜ banzai form migrate --help
Migrate form values from source config to target config

Usage:
  banzai form migrate SOURCE_FORM_CONFIG TARGET_FORM_CONFIG [flags]

➜ banzai form open --help   
Open form defined in the form config file

Usage:
  banzai form open FORM_CONFIG [--port 0] [--browser] [flags]

Flags:
      --browser    open browser form
  -h, --help       help for open
      --port int   port number

➜ banzai form template --help
Execute form template(s) using values from the provided values in the config file

Usage:
  banzai form template FORM_CONFIG [-n TEMPLATE_NAME] [--force] [flags]

Flags:
      --force         overwrite existing files without prompt
  -h, --help          help for template
  -n, --name string   template name

➜ banzai form open ./form-config.yaml --browser
# Single template: banzai form template CONFIG [-n NAME]
➜ banzai form template ./form-config.yaml
```

Example `form-config.yaml`

```yaml
form:
  - description: Credentials are required to support price comparisons of infrastructures
      on different cloud providers
    fields:
      - controlType: checkbox
        default: false
        key: enable_google
        label: enable
        placeholder: Google
      - controlGroupType: google
        key: google
        label: ''
        showIf:
          properties:
            enable_google:
              const: true
      - controlType: checkbox
        default: false
        key: enable_amazon
        label: enable
        placeholder: Amazon
      - controlGroupType: amazon
        key: amazon
        label: ''
        showIf:
          properties:
            enable_amazon:
              const: true
      - controlType: checkbox
        default: false
        key: enable_alibaba
        label: enable
        placeholder: Alibaba
        value: false
      - controlGroupType: alibaba
        key: alibaba
        label: ''
        showIf:
          properties:
            enable_alibaba:
              const: true
      - controlType: checkbox
        default: false
        key: enable_azure
        label: enable
        placeholder: Azure
        value: false
      - controlGroupType: azure
        key: azure
        label: ''
        showIf:
          properties:
            enable_azure:
              const: true
      - controlType: checkbox
        default: false
        key: enable_oracle
        label: enable
        placeholder: Oracle
        value: false
      - controlGroupType: oracle
        key: oracle
        label: ''
        showIf:
          properties:
            enable_oracle:
              const: true
    name: Provider credentials
  - fields:
      - controlType: text
        description: Fully qualified domain name
        key: cloudinfo_fqdn
        label: FQDN
        pattern: ^(([a-zA-Z0-9]|[a-zA-Z0-9][a-zA-Z0-9\-]*[a-zA-Z0-9])\.){2,}([A-Za-z0-9]|[A-Za-z0-9][A-Za-z0-9\-]*[A-Za-z0-9]){2,}$
        placeholder: www.example.com
        required: true
    link: https://github.com/banzaicloud/cloudinfo
    name: Cloud Info
  - fields:
      - controlType: text
        description: Fully qualified domain name
        key: recommender_fqdn
        label: FQDN
        pattern: ^(([a-zA-Z0-9]|[a-zA-Z0-9][a-zA-Z0-9\-]*[a-zA-Z0-9])\.){2,}([A-Za-z0-9]|[A-Za-z0-9][A-Za-z0-9\-]*[A-Za-z0-9]){2,}$
        placeholder: www.example.com
        required: true
    link: https://github.com/banzaicloud/telescopes
    name: Recommender
templates:
  values.yaml: |
    app:
      basePath: /recommender
      cloudInfoAddress: http://{{ .recommender_fqdn }}/cloudinfo/api/v1/
    ingress:
      enabled: true
      hosts:
      - {{ .recommender_fqdn }}/recommender
    {{- if .enable_google }}
    google:
      key: |
        {{ .google.json_key | nindent 4 }}
    {{- end }}

```

Client:

image

image

### CODE REVIEW

File: internal/cli/command/form/open.go [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+
+	port := listener.Addr().(*net.TCPAddr).Port
+	log.Debugf("using port %d\n", port)
+
+	web := packr.New("web", path.Join(".", "web"))
+	http.Handle("/", http.FileServer(web))
+	http.HandleFunc("/api/v1/form", createHandler(options))
+
+	if options.openBrowser {
+		open.Start("http://127.0.0.1:" + strconv.Itoa(port))
```
Comment (sagikazarmark) [inline internal/cli/command/form/open.go]: Use `fmt.Sprintf` instead of `strconv`

File: internal/cli/command/form/web/main.js [hunk 2]
```diff
@@ -0,0 +1,2333 @@
+(window["webpackJsonp"] = window["webpackJsonp"] || []).push([["main"],{
```
Comment (orymate) [inline internal/cli/command/form/web/main.js]: where does this file come from? do we really want to add it to git?

File: internal/cli/command/form/open.go [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+	}
+
+	cmd.Flags().BoolVar(&options.openBrowser, "browser", false, "open browser form")
+	cmd.Flags().IntVar(&options.port, "port", 0 /* find an open port */, "port number")
+
+	return cmd
+}
+
+func runOpen(banzaiCli cli.Cli, options openOptions) {
+	listener, err := net.Listen("tcp", ":"+fmt.Sprintf("%d", options.port))
```
Comment (orymate) [inline internal/cli/command/form/open.go]: Bind on the loopback device (at least by default).

File: internal/cli/command/form/web/main.js (hunk 2 above)
Comment (tothandras) [reply internal/cli/command/form/web/main.js]: We need to decide where should we keep the source code. I'd prefer to have it inside this repository.

Comment (orymate) [reply internal/cli/command/form/web/main.js]: but this is generated code, isn't this?

Comment (tothandras) [reply internal/cli/command/form/web/main.js]: Yes, for now, it is.

File: internal/cli/command/form/open.go [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+	}
+
+	cmd.Flags().BoolVar(&options.openBrowser, "browser", false, "open browser form")
+	cmd.Flags().IntVar(&options.port, "port", 0 /* find an open port */, "port number")
+
+	return cmd
+}
+
+func runOpen(banzaiCli cli.Cli, options openOptions) {
+	listener, err := net.Listen("tcp", "127.0.0.1:"+fmt.Sprintf("%d", options.port))
```
Comment (orymate) [inline internal/cli/command/form/open.go]:
```suggestion
	listener, err := net.Listen("tcp", fmt.Sprintf("127.0.0.1:%d", options.port))
```

File: Makefile [hunk 5]
```diff
@@ -42,10 +42,17 @@ bin/dep-${DEP_VERSION}:
...
+	go get -u github.com/gobuffalo/packr/v2/packr2
+
 .PHONY: vendor
-vendor: bin/dep ## Install dependencies
+vendor: bin/dep bin/packr2 ## Install dependencies
 	bin/dep ensure -v -vendor-only
 
+.PHONY: pre-build ## Pre build bundles of static assets
+pre-build:
+	@cd internal/cli/command/form && packr2
```
Comment (orymate) [inline Makefile]:
```suggestion
	cd internal/cli/command/form && $(abspath bin/packr2)
```

do not hide the command, and do not assume that current gopath/bin is in the path (untested suggestion)

File: Makefile [hunk 6]
```diff
@@ -46,20 +47,38 @@ bin/dep-${DEP_VERSION}:
...
+	curl -L https://github.com/myitcv/gobin/releases/download/v${GOBIN_VERSION}/linux-amd64 > ./bin/gobin-${GOBIN_VERSION} && chmod +x ./bin/gobin-${GOBIN_VERSION}
+endif
+
+bin/packr2: bin/gobin
+	@mkdir -p bin
+	GOBIN=bin/ bin/gobin github.com/gobuffalo/packr/v2/packr2
+
+.PHONY: pre-build ## Pre build bundles of static assets
+pre-build: bin/packr2
+	cd internal/cli/command/form && $(abspath bin/packr2)
```
Comment (sagikazarmark) [inline Makefile]:
```suggestion
	cd internal/cli/command/form
	bin/packr2
```

File: Makefile [hunk 7]
```diff
@@ -28,6 +28,7 @@ GOLANGCI_VERSION = 1.12.2
 GOTESTSUM_VERSION = 0.3.2
 LICENSEI_VERSION = 0.0.7
 GORELEASER_VERSION = 0.98.0
+GOBIN_VERSION = 0.0.4
```
Comment (sagikazarmark) [inline Makefile]:
```suggestion
GOBIN_VERSION = 0.0.4
PACKR_VERSION = 2.0.1
```

File: Makefile [hunk 8]
```diff
@@ -46,20 +47,38 @@ bin/dep-${DEP_VERSION}:
...
+bin/gobin-${GOBIN_VERSION}:
+	@mkdir -p bin
+ifeq (${OS}, Darwin)
+	curl -L https://github.com/myitcv/gobin/releases/download/v${GOBIN_VERSION}/darwin-amd64 > ./bin/gobin-${GOBIN_VERSION} && chmod +x ./bin/gobin-${GOBIN_VERSION}
+endif
+ifeq (${OS}, Linux)
+	curl -L https://github.com/myitcv/gobin/releases/download/v${GOBIN_VERSION}/linux-amd64 > ./bin/gobin-${GOBIN_VERSION} && chmod +x ./bin/gobin-${GOBIN_VERSION}
+endif
+
+bin/packr2: bin/gobin
```
Comment (sagikazarmark) [inline Makefile]:
```suggestion
bin/packr2: bin/packr2-${PACKR_VERSION}
	@ln -sf packr2-${PACKR_VERSION} bin/packr2
bin/packr2-${PACKR_VERSION}: bin/gobin
```

File: Makefile [hunk 9]
```diff
@@ -46,20 +47,38 @@ bin/dep-${DEP_VERSION}:
...
+	@mkdir -p bin
+	GOBIN=bin/ bin/gobin github.com/gobuffalo/packr/v2/packr2
```
Comment (sagikazarmark) [inline Makefile]:
```suggestion
	GOBIN=bin/ bin/gobin github.com/gobuffalo/packr/v2/packr2@${PACKR_VERSION}
```

File: internal/cli/command/form/template.go [hunk 10]
```diff
@@ -0,0 +1,122 @@
...
+		Long:  "Execute form template(s) using values from the provided values in the config file",
+		Args:  cobra.ExactArgs(1),
+		Run: func(cmd *cobra.Command, args []string) {
+			options.configFile = args[0]
+			runExecuteTemplate(banzaiCli, options)
+		},
+	}
+
+	cmd.Flags().StringVarP(&options.name, "name", "n", "", "template name")
+	cmd.Flags().BoolVar(&options.force, "force", false, "overwrite existing files without prompt")
```
Comment (sagikazarmark) [inline internal/cli/command/form/template.go]:
```suggestion
	cmd.Flags().BoolVar(&options.force, "force", false, "overwrite existing files")
```

File: Makefile [hunk 11]
```diff
@@ -64,7 +64,8 @@ bin/packr2: bin/gobin
 
 .PHONY: pre-build ## Pre build bundles of static assets
 pre-build: bin/packr2
-	cd internal/cli/command/form && $(abspath bin/packr2)
+	cd internal/cli/command/form
+	bin/packr2
```
Comment (orymate) [inline Makefile]: noes, revert pls

Comment (sagikazarmark) [reply Makefile]: Sry, my bad.

File: internal/cli/command/form/web/main.js (hunk 2 above)
Comment (tothandras) [reply internal/cli/command/form/web/main.js]: Added the client source as well.

**Pull request merged**